include/seahorn/Expr/EvalModel.hh

```diff
@@ -0,0 +1,302 @@
+#pragma once
+
+#include "seahorn/Expr/EvalUtils.hh"
+#include "seahorn/Expr/ExprCore.hh"
+#include "seahorn/Expr/ExprGmp.hh"
+#include "seahorn/Expr/ExprOpBind.hh"
+#include "seahorn/Expr/ExprOpBinder.hh"
+
+#include <math.h>
+#include <stdlib.h>
+
+namespace expr {
+namespace eval {
+
+/**
+ * Bit vector structure. Holds the number(as type T) and its width
+ *
+ * \note bools are stored as 1/0 with a width of 1
+ */
+template <typename T> struct BvNum {
+  T num = 0;
+  unsigned width = 0;
+
+  /// \brief empty constructor
+  BvNum() : num(0), width(0) {}
+
+  /// \brief bool constructor.
+  BvNum(bool a) : num(a), width(1) {}
+
+  BvNum(T numArg, unsigned widthArg) : num(numArg), width(widthArg) {}
+
+  bool operator<(const BvNum<T> &bnum) const { return num < bnum.num; }
+
+  bool operator==(const BvNum<T> &bnum) const {
+    return (num == bnum.num) && (width == bnum.width);
+  }
+};
+
+/**
+ * Holds the data for an array.
+ *
+ * In EvalModel, the array expressions are mapped to this struct
+ *
+ */
+template <typename T> struct Arr {
+  bool m_hasDefaultVal = false;
+  BvNum<T> m_defaultVal;
+
+  unsigned m_valWidth = 0;
+
+  std::map<BvNum<T>, BvNum<T>> m_map; ///< index, value
+
+  Arr(BvNum<T> defaultVal)
+      : m_hasDefaultVal(true), m_defaultVal(defaultVal),
+        m_valWidth(defaultVal.width) {}
+  Arr(unsigned valWidth) : m_hasDefaultVal(false), m_valWidth(valWidth) {}
+};
+
+/**
+ * Holds the data for a lambda. Maps bound vars to their value
+ *
+ * The purpose of this is to keep bound vars within the scope of the lambda. The
+ * same bound var expression could have different values in different lambdas
+ *
```

Comment:
In general, lambda is a function. What does that mean for `bound var` to have different values?
Say I have lambda expressions of this form : `[](bvnum x) { return x + 1; }`, where `x` is the bound variable.
I then get stuck trying to parse the comment and apply it to the lambda above.

Reply:
I might need to change the wording, I meant the expressions itself. Is it possible that you might have a bound variable expression that is used in more than one lambda? Obviously the value of it would be different depending on where its used but the Expr is the same

Reply:
This makes sense. Bound variables are arguments and each lambda can use them in their own way. 

Seems like you only consider lambdas that can be written as a table. This is ok for now